java/org/apache/catalina/webresources/AbstractSingleArchiveResourceSet.java

```diff
@@ -63,29 +63,32 @@ public AbstractSingleArchiveResourceSet(WebResourceRoot root, String webAppMount
 
     @Override
     protected HashMap<String,JarEntry> getArchiveEntries(boolean single) {
-        synchronized (archiveLock) {
-            if (archiveEntries == null && !single) {
-                JarFile jarFile = null;
-                archiveEntries = new HashMap<>();
-                try {
-                    jarFile = openJarFile();
-                    Enumeration<JarEntry> entries = jarFile.entries();
-                    while (entries.hasMoreElements()) {
-                        JarEntry entry = entries.nextElement();
-                        archiveEntries.put(entry.getName(), entry);
-                    }
-                } catch (IOException ioe) {
-                    // Should never happen
-                    archiveEntries = null;
-                    throw new IllegalStateException(ioe);
-                } finally {
-                    if (jarFile != null) {
-                        closeJarFile();
+        if (archiveEntries == null && !single) {
+            synchronized (archiveLock) {
+                if (archiveEntries == null) {
+                    JarFile jarFile = null;
+                    try {
+                        HashMap<String,JarEntry> contents = new HashMap<>();
+                        jarFile = openJarFile();
+                        Enumeration<JarEntry> entries = jarFile.entries();
+                        while (entries.hasMoreElements()) {
+                            JarEntry entry = entries.nextElement();
+                            contents.put(entry.getName(), entry);
+                        }
+                        archiveEntries = contents;
+                    } catch (IOException ioe) {
+                        // Should never happen
+                        archiveEntries = null;
+                        throw new IllegalStateException(ioe);
+                    } finally {
+                        if (jarFile != null) {
+                            closeJarFile();
+                        }
                     }
                 }
             }
-            return archiveEntries;
         }
+        return archiveEntries;
```

Comment:
Given the bugs we were seeing, should we also make `archiveEntries` `volatile` this time?

Reply:
Sure, I can add that

Reply:
I'm looking at the `volatile` thing...the variable is up in a superclass.  I remember that keyword causing a performance issue in log4j such that we removed it from the category class there as a patch.  Is that a concern here?  Would it buy us anything in this version of the patched code?

Reply:
Since it's only written once, I don't think perf would be a problem.  Should still be faster than obtaining a lock every time.